src/lv_misc/lv_txt_ap.c

```diff
@@ -0,0 +1,263 @@
+
+		i++;
+    }
+
+    return chars_cnt + 1;
+}
+
+
+void lv_txt_ap_proc(const char * txt,char * txt_out){
+    uint32_t txt_length=0;
+    uint32_t index_current,idx_next,idx_previous,i,j;
+    uint32_t * ch_enc;
+    char * txt_out_temp;
+
+    txt_length = lv_txt_get_encoded_length(txt);
+
+    ch_enc = (uint32_t *)lv_mem_alloc(sizeof(uint32_t) * txt_length);
+
+    i=0;
+    j=0;
+    while(j<txt_length)
+    	ch_enc[j++] = lv_txt_encoded_next(txt, &i);
+
+    i=0;
+    idx_previous = LV_UNDEF_ARABIC_PERSIAN_CHARS;
+    while(i<txt_length){
+        index_current = lv_ap_get_char_index(ch_enc[i]);
+        idx_next = lv_ap_get_char_index(ch_enc[i+1]);
+
+        if(index_current == LV_UNDEF_ARABIC_PERSIAN_CHARS)
+        {
+          i++;
+          idx_previous = LV_UNDEF_ARABIC_PERSIAN_CHARS;
+          continue;
+        }
+
+        uint8_t conjunction_to_previuse = (i==0 || idx_previous == LV_UNDEF_ARABIC_PERSIAN_CHARS) ? 0 : ap_chars_map[idx_previous].ap_chars_conjunction.conj_to_next;
+        uint8_t conjunction_to_next = ((i==txt_length-1) || idx_next == LV_UNDEF_ARABIC_PERSIAN_CHARS) ? 0 : ap_chars_map[idx_next].ap_chars_conjunction.conj_to_previous;
+
+        if(conjunction_to_previuse && conjunction_to_next)
+          ch_enc[i] = ap_chars_map[index_current].char_end_form+ap_chars_map[index_current].char_middle_form_offset;
+        else if(!conjunction_to_previuse && conjunction_to_next)
+          ch_enc[i] = ap_chars_map[index_current].char_end_form+ap_chars_map[index_current].char_begining_form_offset;
+        else if(conjunction_to_previuse && !conjunction_to_next)
+          ch_enc[i] = ap_chars_map[index_current].char_end_form;
+        else
+          ch_enc[i] = ap_chars_map[index_current].char_end_form+ap_chars_map[index_current].char_isolated_form_offset;
+        idx_previous = index_current;
+        i++;
+    }
+
+#if LV_USE_REVERSE_ARABIC_PERSIAN_CHARS == 1
+    lv_ap_normalize_chars(ch_enc,txt_length);
+#endif
+
+    txt_out_temp = txt_out;
+    i=0;
+    while(i<txt_length){
+      if((*ch_enc) <=0x7F){
+          *(txt_out_temp++) = *ch_enc;
+      }else if((*ch_enc) <=0x7FFF){
+          *(txt_out_temp++) = 0xC0 | ((*ch_enc>>6) & 0x3F);
+          *(txt_out_temp++) = 0x80 | (*ch_enc & 0x3F);
+      }else {
+        *(txt_out_temp++) = 0xE0 | ((*ch_enc>>12) & 0x3F);
+        *(txt_out_temp++) = 0x80 | ((*ch_enc>>6) & 0x3F);
+        *(txt_out_temp++) = 0x80 | (*ch_enc & 0x3F);
+      }
+      ch_enc++;
+      i++;
+  }
+  *(txt_out_temp++)=0;
+  lv_mem_free(ch_enc);
+}
+/**********************
+*   STATIC FUNCTIONS
+**********************/
+
+static uint32_t lv_ap_get_char_index(uint16_t c){
+    for(uint8_t i=0;ap_chars_map[i].char_end_form;i++){
+        if(c == (ap_chars_map[i].char_offset + LV_AP_ALPHABET_BASE_CODE))
+          return i;
+    }
+    return LV_UNDEF_ARABIC_PERSIAN_CHARS;
+}
+
+#if LV_USE_REVERSE_ARABIC_PERSIAN_CHARS == 1
+static void lv_ap_normalize_chars(uint32_t * str,uint16_t count){
```

Comment:
This needs some clarification.
lvgl already has bidi functionality to represent mixed RTL and LTR text (like numbers and text, neutral characters such as punctuation marks etc.) 
Isn't `lv_ap_normalize_chars` re-implementing the same functionality?

Reply:
Yes, You are right. It must be removed totally.